src/Build/BackEnd/Node/RarNode.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.CodeDom;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Internal;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Execution
+{
+    public sealed class RarNode : INode
+    {
+        /// <summary>
+        /// The amount of time to wait for the client to connect to the host.
+        /// </summary>
+        private const int ClientConnectTimeout = 60000;
+
+        public NodeEngineShutdownReason Run(bool nodeReuse, bool lowPriority, out Exception shutdownException)
+        {
+            shutdownException = null;
+            using CancellationTokenSource cts = new CancellationTokenSource();
+            string pipeName = CommunicationsUtilities.GetRarPipeName(nodeReuse, lowPriority);
+            IRarController controller = GetController(pipeName);
+
+            Task<int> rarTask = controller.StartAsync(cts.Token);
+
+            Handshake handshake = NodeProviderOutOfProc.GetHandshake(enableNodeReuse: nodeReuse,
+                                                                     enableLowPriority: lowPriority, specialNode: true);
+            Task<NodeEngineShutdownReason> msBuildShutdown = RunShutdownCheckAsync(handshake, cts.Token);
+
+            // Wait for any of these task to finish:
+            // - rarTask can timeout (default is 15 mins)
+            // - msBuildShutdown ends when it recieves command to shutdown
+            int index = Task.WaitAny(msBuildShutdown, rarTask);
+            cts.Cancel();
+
+            if (index == 0)
+            {
+                // We know that this task is completed so we can get Result without worring about waiting for it
+                return msBuildShutdown.Result;
+            }
+            else
+            {
+                // RAR task can only exit with connection error or timeout
+                return NodeEngineShutdownReason.ConnectionFailed;
+            }
+        }
+
+        private static IRarController GetController(string pipeName)
+        {
+            const string rarControllerName = "Microsoft.Build.Tasks.ResolveAssemblyReferences.Server.RarController, Microsoft.Build.Tasks.Core";
+            Type rarControllerType = Type.GetType(rarControllerName);
+
+            Func<string, int?, int?, int, bool, NamedPipeServerStream> streamFactory = NamedPipeUtil.CreateNamedPipeServer;
+            IRarController controller = (IRarController)Activator.CreateInstance(rarControllerType, pipeName, streamFactory, null);
```

Comment:
Is it legal to do this?
```suggestion
            IRarController controller = (IRarController)Activator.CreateInstance(rarControllerType, pipeName, NamedPipeUtil.CreateNamedPipeServer, null);
```

Reply:
Without cast to Func it is not. If I just pass the method, it is considered as method group, which is not object. Of course, I could cast it inside the CreateInstance call, but this way it looks cleaner for me.

Reply:
Related to this https://github.com/dotnet/msbuild/pull/5613/files#r478309866. TLDR: netstandard2.0 doesn't have all API which are used in NamedPipeUtil